wallet/src/Pos/Wallet/Web/Methods/History.hs

```diff
@@ -116,88 +116,90 @@ getHistoryLimited
     -> Maybe Word
     -> Maybe Word
     -> m ([CTx], Word)
-getHistoryLimited mCWalId mAccId mAddrId mSkip mLimit =
-    first applySkipLimit <$> getHistory mCWalId mAccId mAddrId
+getHistoryLimited mCWalId mAccId mAddrId mSkip mLimit = do
+    (cWalId, accIds) <- case (mCWalId, mAccId) of
+        (Nothing, Nothing)      -> throwM errorSpecifySomething
+        (Just _, Just _)        -> throwM errorDontSpecifyBoth
+        (Just cWalId', Nothing) -> do
+            accIds' <- getWalletAccountIds cWalId'
+            pure (cWalId', accIds')
+        (Nothing, Just accId)   -> pure (aiWId accId, [accId])
+    (unsortedThs, n) <- getHistory cWalId accIds mAddrId
+    let sortedTxh = sortByTime (Map.elems unsortedThs)
+    pure (applySkipLimit sortedTxh, n)
   where
+    sortByTime :: [(CTx, POSIXTime)] -> [CTx]
+    sortByTime thsWTime =
+        -- TODO: if we use a (lazy) heap sort here, we can get the
+        -- first n values of the m sorted elements in O(m + n log m)
+        map fst $ sortWith (Down . snd) thsWTime
     applySkipLimit = take limit . drop skip
     limit = (fromIntegral $ fromMaybe defaultLimit mLimit)
     skip = (fromIntegral $ fromMaybe defaultSkip mSkip)
     defaultLimit = 100
     defaultSkip = 0
+    errorSpecifySomething = RequestError $
+        "Please specify either walletId or accountId"
+    errorDontSpecifyBoth = RequestError $
+        "Please do not specify both walletId and accountId at the same time"
 
 addHistoryTx
     :: MonadWalletWebMode m
     => CId Wal
     -> TxHistoryEntry
     -> m ()
-addHistoryTx cWalId = addHistoryTxs cWalId . one
+addHistoryTx cWalId = addHistoryTxs cWalId . txHistoryListToMap . one
 
 -- This functions is helper to do @addHistoryTx@ for
 -- all txs from mempool as one Acidic transaction.
 addHistoryTxs
     :: MonadWalletWebMode m
     => CId Wal
-    -> [TxHistoryEntry]
+    -> Map TxId TxHistoryEntry
     -> m ()
 addHistoryTxs cWalId historyEntries = do
     metas <- mapM toMeta historyEntries
-    let cIds = map (encodeCType . _thTxId) historyEntries
-    addOnlyNewTxMetas cWalId (zip cIds metas)
+    addOnlyNewTxMetas cWalId metas
   where
     toMeta THEntry {..} = CTxMeta <$> case _thTimestamp of
         Nothing -> liftIO getPOSIXTime
         Just ts -> pure $ timestampToPosix ts
 
 constructCTx
     :: MonadWalletWebMode m
-    => (CId Wal, Maybe [CWAddressMeta])
+    => CId Wal
+    -> Set (CId Addr)
+    -> ChainDifficulty
     -> TxHistoryEntry
-    -> m CTx
-constructCTx (cWalId, walAddrMetasMB) wtx@THEntry{..} = do
-    walAddrMetas <- maybe (getWalletAddrMetas Ever cWalId) pure walAddrMetasMB
+    -> m (CTx, POSIXTime)
```

Comment:
Do we need to return `POSIXTime` explicitly here? `CTx` already contains the same thing..